C codegen: nested `or` blocks emit invalid C (`*(T*) _t.data = ;`) when outer call result is discarded

## V version
`V 0.5.1 1510663` on Windows 11, MSVC backend (`-cc msvc`).

## Description
When a function returning `!T` (T non-void) is called as a **statement** (return value discarded) with an `or { ... }` block whose **tail expression is itself an `or` block** on a `!void` call, the C generator emits an assignment with an empty right-hand side:

```c
*(T*) _t.data = ;
```

MSVC rejects this with `error C2059: syntax error: ';'`.

## Minimal reproduction

```v
module main

struct Item {
	id int
}

fn get_item() !Item {
	return Item{id: 1}
}

fn do_void() ! {
	return error('boom')
}

fn main() {
	get_item() or {
		do_void() or {
			println('inner: ${err}')
		}
	}
}
```

Build:
```
v -cc msvc -keepc repro.v
```

Result:
```
repro.exe.tmp.c(8373): error C2059: syntax error: ';'
builder error: msvc error
```

## Generated C (excerpt)

```c
VV_LOC void main__main(void) {
    _result_main__Item _t1 = main__get_item();
    if (_t1.is_error) {
        IError err = _t1.err;
        _result_void _t2 = main__do_void();
        if (_t2.is_error) {
            IError err = _t2.err;
            builtin__println(...);
        ;
        }

        *(main__Item*) _t1.data = ;   // <-- invalid C, empty RHS
    }

 ;
}
```

The compiler appears to treat the inner `or` block (a `!void` expression used as a statement) as a value expression yielding `Item`, then emits the standard "store the or-block's value into `_t1.data`" assignment with no RHS because there is no value.

## Expected behavior

Either:
1. No assignment should be emitted at all, since the outer call's result is discarded (the call is used as a statement); the `or` block exists only for side-effecting error handling, or
2. A V-level error/warning should reject the program if codegen cannot lower it.

## Workarounds

Any of these compile and behave correctly:

- Use the `if x := call() { } else { }` form on the outer call instead of `or { ... }`.
- Extract the inner `do_void() or { ... }` into a helper function so the outer `or` block does not end with a nested `or` expression.
- Add a non-`or` trailing statement inside the outer `or` block (e.g. `return`, `continue`, or another non-`or` call).

## Notes

- Reproduces on all three C backends I tested (Windows 11):
  - `-cc msvc` → `error C2059: syntax error: ';'`
  - `-cc tcc` → `error: identifier expected`
  - `-cc gcc` → `error: expected expression before ';' token`
- Triggered specifically by **nesting** — a single-level `or { do_void() or { ... } }` is the smallest shape that hits it. A non-`or` tail in the outer block (e.g. `or { do_void() or { ... }; return }`) avoids the bug.


> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -11922,6 +11922,15 @@ fn (mut g Gen) or_block_on_value(var_name string, or_block ast.OrExpr, return_ty
 				else {}
 			}
 		}
+		// `call() or { ... }` whose call returns `!void`/`?void` has already
+		// consumed the option/result via its own or-block; the effective
+		// value of the last expression is void, so do not try to write
+		// `*(T*) tmp.data = <void>` (see issue #27012).
+		if last_expr_stmt.expr is ast.CallExpr
+			&& last_expr_stmt.expr.or_block.kind != .absent
+			&& last_expr_typ.clear_option_and_result() == ast.void_type {
+			last_expr_typ = ast.void_type
+		}
 		last_expr_produces_value = last_expr_typ != ast.void_type
 	}
 	if last_expr_produces_value {
@@ -12030,6 +12039,15 @@ fn (mut g Gen) or_block(var_name string, or_block ast.OrExpr, return_type ast.Ty
 					}
 					else {}
 				}
+			}
+			// `call() or { ... }` whose call returns `!void`/`?void` has already
+			// consumed the option/result via its own or-block; the effective
+			// value of the last expression is void, so do not try to write
+			// `*(T*) tmp.data = <void>` (see issue #27012).
+			if last_expr_stmt.expr is ast.CallExpr
+				&& last_expr_stmt.expr.or_block.kind != .absent
+				&& last_expr_typ.clear_option_and_result() == ast.void_type {
+				last_expr_typ = ast.void_type
 			}
 			last_expr_produces_value = last_expr_typ != ast.void_type
 		}

--- a/vlib/v/tests/options/nested_or_block_with_void_result_test.v
+++ b/vlib/v/tests/options/nested_or_block_with_void_result_test.v
@@ -1,0 +1,45 @@
+// Regression test for https://github.com/vlang/v/issues/27012
+// A nested `or { call_returning_result_void() or { ... } }` used as a
+// statement (outer call result discarded) used to emit invalid C of the
+// form `*(T*) _t.data = ;` and fail to compile.
+
+struct Item {
+	id int
+}
+
+fn get_item() !Item {
+	return Item{id: 1}
+}
+
+fn get_item_err() !Item {
+	return error('outer')
+}
+
+fn do_void() ! {
+	return error('inner')
+}
+
+fn test_nested_or_block_with_void_result_inner_succeeds() {
+	mut traces := []string{}
+	get_item() or {
+		traces << 'outer: ${err}'
+		do_void() or {
+			traces << 'inner: ${err}'
+		}
+	}
+	// Outer call succeeded -> or-block is never entered.
+	assert traces.len == 0
+}
+
+fn test_nested_or_block_with_void_result_outer_errors() {
+	mut traces := []string{}
+	get_item_err() or {
+		traces << 'outer: ${err}'
+		do_void() or {
+			traces << 'inner: ${err}'
+		}
+	}
+	assert traces.len == 2
+	assert traces[0] == 'outer: outer'
+	assert traces[1] == 'inner: inner'
+}

